File: digital_signature.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/rsa.h>
#include <openssl/pem.h>
#include <openssl/sha.h>
#include <openssl/err.h>

// Function to generate key pair
void generate_key_pair() {
    RSA *rsa = RSA_new();
    BIGNUM *bne = BN_new();
    BN_set_word(bne, RSA_F4);
    RSA_generate_key_ex(rsa, 2048, bne, NULL);

    // Save private key
    FILE *private_key_file = fopen("private.pem", "wb");
    PEM_write_RSAPrivateKey(private_key_file, rsa, NULL, NULL, 0, NULL, NULL);
    fclose(private_key_file);

    // Save public key
    FILE *public_key_file = fopen("public.pem", "wb");
    PEM_write_RSAPublicKey(public_key_file, rsa);
    fclose(public_key_file);

    RSA_free(rsa);
    BN_free(bne);
}

// Function to create message digest (hash)
void create_digest(const unsigned char *message, size_t message_len, unsigned char *digest) {
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, message, message_len);
    SHA256_Final(digest, &sha256);
}

// Function to sign a message
unsigned char* sign_message(const char *message, size_t *signature_len) {
    // Read private key
    FILE *private_key_file = fopen("private.pem", "rb");
    RSA *rsa = PEM_read_RSAPrivateKey(private_key_file, NULL, NULL, NULL);
    fclose(private_key_file);

    // Create digest
    unsigned char digest[SHA256_DIGEST_LENGTH];
    create_digest((unsigned char*)message, strlen(message), digest);

    // Sign the digest
    unsigned char *signature = malloc(RSA_size(rsa));
    RSA_private_encrypt(SHA256_DIGEST_LENGTH, digest, signature, rsa, RSA_PKCS1_PADDING);
    *signature_len = RSA_size(rsa);

    RSA_free(rsa);
    return signature;
}

// Function to verify signature
int verify_signature(const char *message, const unsigned char *signature, size_t signature_len) {
    // Read public key
    FILE *public_key_file = fopen("public.pem", "rb");
    RSA *rsa = PEM_read_RSAPublicKey(public_key_file, NULL, NULL, NULL);
    fclose(public_key_file);

    // Decrypt signature
    unsigned char decrypted[SHA256_DIGEST_LENGTH];
    RSA_public_decrypt(signature_len, signature, decrypted, rsa, RSA_PKCS1_PADDING);

    // Create digest of original message
    unsigned char digest[SHA256_DIGEST_LENGTH];
    create_digest((unsigned char*)message, strlen(message), digest);

    // Compare decrypted signature with digest
    int result = memcmp(decrypted, digest, SHA256_DIGEST_LENGTH) == 0;

    RSA_free(rsa);
    return result;
}

int main() {
    // Initialize OpenSSL
    OpenSSL_add_all_algorithms();
    ERR_load_crypto_strings();

    // Generate key pair
    printf("Generating key pair...\n");
    generate_key_pair();
    printf("Keys generated successfully!\n\n");

    // Test message
    const char *message = "Hello, this is a test message!";
    printf("Original message: %s\n", message);

    // Sign message
    size_t signature_len;
    unsigned char *signature = sign_message(message, &signature_len);
    printf("Message signed successfully!\n");

    // Verify signature
    if (verify_signature(message, signature, signature_len)) {
        printf("Signature verification successful!\n");
    } else {
        printf("Signature verification failed!\n");
    }

    // Test with tampered message
    const char *tampered_message = "Hello, this is a modified message!";
    if (verify_signature(tampered_message, signature, signature_len)) {
        printf("Warning: Tampered message verified!\n");
    } else {
        printf("Security working: Tampered message detected!\n");
    }

    // Cleanup
    free(signature);
    EVP_cleanup();
    ERR_free_strings();

    return 0;
}

File: digital_signature.py
from Crypto.PublicKey import RSA
from Crypto.Signature import pkcs1_15
from Crypto.Hash import SHA256

class DigitalSignature:
    def __init__(self):
        self.key_pair = None
        self.public_key = None
        self.private_key = None

    def generate_key_pair(self):
        # Tạo cặp khóa mới
        self.key_pair = RSA.generate(2048)
        self.private_key = self.key_pair
        self.public_key = self.key_pair.publickey()
        
        # Lưu khóa
        with open("signature_private.pem", "wb") as f:
            f.write(self.private_key.export_key('PEM'))
        with open("signature_public.pem", "wb") as f:
            f.write(self.public_key.export_key('PEM'))

    def sign_message(self, message: str) -> bytes:
        # Tạo hash của message
        h = SHA256.new(message.encode('utf-8'))
        
        # Ký hash bằng private key
        signature = pkcs1_15.new(self.private_key).sign(h)
        return signature

    def verify_signature(self, message: str, signature: bytes) -> bool:
        # Tạo hash của message
        h = SHA256.new(message.encode('utf-8'))
        
        try:
            # Xác thực chữ ký bằng public key
            pkcs1_15.new(self.public_key).verify(h, signature)
            return True
        except (ValueError, TypeError):
            return False

def main():
    try:
        # Khởi tạo đối tượng DigitalSignature
        ds = DigitalSignature()
        
        # Tạo cặp khóa
        print("Đang tạo cặp khóa cho chữ ký số...")
        ds.generate_key_pair()
        print("Đã tạo cặp khóa thành công!")

        # Tin nhắn cần ký
        message = "Đây là tin nhắn cần được ký số để đảm bảo tính xác thực."
        print(f"\nTin nhắn gốc: {message}")

        # Ký tin nhắn
        print("\nĐang ký tin nhắn...")
        signature = ds.sign_message(message)
        print("Đã ký tin nhắn thành công!")

        # Xác thực chữ ký
        print("\nĐang xác thực chữ ký...")
        if ds.verify_signature(message, signature):
            print("Xác thực chữ ký thành công!")
        else:
            print("Xác thực chữ ký thất bại!")

        # Thử nghiệm với tin nhắn bị thay đổi
        tampered_message = message + " [Đã bị thay đổi]"
        print(f"\nThử nghiệm với tin nhắn bị thay đổi: {tampered_message}")
        if ds.verify_signature(tampered_message, signature):
            print("CẢNH BÁO: Xác thực thành công với tin nhắn bị thay đổi!")
        else:
            print("Bảo mật hoạt động tốt: Phát hiện tin nhắn bị thay đổi!")

    except Exception as e:
        print(f"Có lỗi xảy ra: {str(e)}")

if __name__ == "__main__":
    main()
